### PR TITLE
Use OPC subscriptions for each publishing interval, remove model compiler, update readme, and various more

### PR DESCRIPTION
**Pull request by hansgschossmann**
This merge will not be the final one, since there is a blocking issue for error free operation in docker scenarios, which will be resolved with a fix in the OPC UA stack most likely with [issue #218](https://github.com/OPCFoundation/UA-.NETStandardLibrary/issues/218). The current stack only takes the first IPv4 and IPv6 addresses returned by the OS into account. This could lead to non connectivity if the target OPC UA server is only reachable via other addresses.

The PR brings in:
- Establish a separate subscription for each publishing interval to a certain endpoint URL
- Remove the model compiler and create the address space programmatically.
- Update the readme.md with the current information
- Add protection of shared data structures
- Change PublishNode and UnpublishNode to support new application architecture
- Use string NodeId's in address space and remove the integer NodeId's
- Change Dockerfile and add a SSH only Dockerfile for SSH debugging in a docker container

### CODE REVIEW

File: src/IotHubMessaging.cs [hunk 1]
```diff
@@ -324,8 +345,6 @@ public void Enqueue(string json)
                 Trace(Utils.TraceMasks.OperationDetail, $"Retart timer to send data to IoTHub in {_defaultSendIntervalSeconds} second(s).");
```
Comment (mregen) [inline src/IotHubMessaging.cs]: typo 'restart' #Resolved

File: src/Namespaces.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+{
+    /// <summary>
+    /// Defines constants for namespaces used by the application.
+    /// </summary>
+    public static partial class Namespaces
+    {
+        /// <summary>
+        /// The namespace for the nodes provided by the publisher server.
+        /// </summary>
+        public const string PublisherApplications = "http://opcfoundation.org/UA/Publisher/";
```
Comment (mregen) [inline src/Namespaces.cs]: Should the namespace URI be rather microsoft.com? #Resolved

File: src/OpcSession.cs [hunk 3]
```diff
@@ -205,184 +264,322 @@ public OpcSession(Uri endpointUri, uint sessionTimeout)
...
         }
 
+        /// <summary>
+        /// Disconnects a session and removes all subscriptions on it and marks all nodes on those subscriptions
+        /// as unmonitored.
+        /// </summary>
+        /// <returns></returns>
         public async Task Disconnect()
         {
             _opcSessionSemaphore.Wait();
```
Comment (mregen) [inline src/OpcSession.cs]: better use await? #Resolved

File: src/OpcSession.cs [hunk 4]
```diff
@@ -205,184 +264,322 @@ public OpcSession(Uri endpointUri, uint sessionTimeout)
...
+                    opcSubscription.OpcMonitoredItems.Add(opcMonitoredItem);
+                    Trace($"AddNodeForMonitoring: Added item with nodeId '{nodeId.ToString()}' for monitoring.");
                 }
-
             }
             finally
             {
                 _opcSessionSemaphore.Release();
+                // Start publishing.
+                Task.Run(async () => await ConnectAndMonitor());
```
Comment (mregen) [inline src/OpcSession.cs]: connect always even on error condition?

File: src/OpcSession.cs [hunk 5]
```diff
@@ -205,184 +264,322 @@ public OpcSession(Uri endpointUri, uint sessionTimeout)
...
             }
         }
 
+        /// <summary>
+        /// Shutsdown all connected sessions.
+        /// </summary>
+        /// <returns></returns>
         public async Task Shutdown()
         {
             _opcSessionSemaphore.Wait();
```
Comment (mregen) [inline src/OpcSession.cs]: I'd avoid blocking calls wherever possible, e.g. await semaphore.WaitAsync() instead of Wait #Resolved

**Review by mregen — APPROVED** [review]
First walkthrough...
Overall seems good.
A little concerned on the Wait/async mix when using semaphores.

File: Dockerfile [hunk 6]
```diff
@@ -1,10 +1,10 @@
...
 
-WORKDIR /build
+WORKDIR /build/src
 RUN dotnet restore
-RUN dotnet publish -c Release -o out
+RUN dotnet publish --configuration Release --output /build/out
 
-WORKDIR /build/out
-ENTRYPOINT ["dotnet", "Opc.Ua.Publisher.dll"]
+WORKDIR /docker
```
Comment (barnstee) [inline Dockerfile:9]: Not sure this will work as the working directory needs to be the publishing directory

File: README.md [hunk 7]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
 
 This application, apart from including an OPC UA *client* for connecting to existing OPC UA servers you have on your network, also includes an OPC UA *server* on port 62222 that can be used to manage what gets published.
 
+The application is implemented using .NET Core technology and is able to run on the platforms supported by .NET Core.
+
+Publisher implements a retry logic to establish connections to endpoints which have not responded to a certain number of keep alive requests, if the OPC UA server on this endpoint had a power outage.
```
Comment (barnstee) [inline README.md]: ... for example if the OPC UA server...

File: README.md [hunk 8]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+
+For each distinct publishing interval to a OPC UA server it creates a separate subscription over which all nodes with this publishing interval are updated.
```
Comment (barnstee) [inline README.md]: an OPC

File: README.md [hunk 9]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+
+Publisher supports aggregation of the data sent to IoTHub, to reduce network load. This batching is sending a packet to IoTHub only if the configured package size is reached.
```
Comment (barnstee) [inline README.md]: I wouldn't call it aggregation, as this is something else on OPC UA. Call it batching throughout.

File: README.md [hunk 10]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+Open the OpcPublisher.sln project with Visual Studio 2017 and build the solution by hitting F7.
+
+## As Docker container
+
+Depending if you use Docker Linux or Docker Windows containers, there are different configuration files (Dockerfile or Dockerfile.Windows) to use for building the container.
+From the root of the repository, in a console, type:
+
+    docker build -f <docker-configfile-to-use> -t <your-container-name> .
+
+The `-f` option for `docker build` is optional and the default is to use Dockerfile. Docker also support building directly from a git repository, which means you also could build a Linux container by:
```
Comment (barnstee) [inline README.md]: can, not could

File: README.md [hunk 11]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+                {
+                    // the identifier specifies the NamespaceUri and the node identifier in XML notation as specified in Part 6 of the OPC UA specification in the XML Mapping section.
+                    "ExpandedNodeId": "nsu=http://opcfoundation.org/UA/;i=2258"
+                }
+            ]
+        },
+
+
+        {
+            // Publisher will request the server at EndpointUrl to sample the node with the OPC sampling interval specified on command line (or the default value: OPC publishing interval).
```
Comment (barnstee) [inline README.md]: I wouldn't put in the comments on every node.

File: README.md [hunk 12]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+The following options are supported:
+
+      --pf, --publishfile=VALUE
+                             the filename to configure the nodes to publish.
+                               Default: '/docker/publishednodes.json'
+      --sd, --shopfloordomain=VALUE
+                             the domain of the shopfloor. if specified this
+                               domain is appended (delimited by a ':' to the '
+                               ApplicationURI' property when telemetry is
+                               ingested to IoTHub.
```
Comment (barnstee) [inline README.md]: sent, not ingested

File: README.md [hunk 13]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+                               (allowed values: Directory, X509Store)
+                               Default: X509Store
+      --dp, --devicecertstorepath=VALUE
+                             the path of the iot device cert store
+                               Default Default (depends on store type):
+                               X509Store: 'IoTHub'
+                               Directory: 'CertificateStores/IoTHub'
+      -h, --help                 show this message and exit
+
+There are a couple of environment variables which could be used to control the application:
```
Comment (barnstee) [inline README.md]: can, not could

File: README.md [hunk 14]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+_GW_PNFP: sets the filename of the publishing configuration file
+
+Command line arguments overrule environment variable settings.
+
+Typically you specify the IoTHub owner connectionstring only on the first start of the application. The connectionstring will be encrypted and stored in the platforms certificiate store.
+On subsequent calls it will be read from there and reused. If you specify the connectionstring on each start, the device which is created for the application in the IoTHub device registry will be removed and recreated each time.
+
+
+## Native on Windows
+Open the OpcPublisher.sln project with Visual Studio 2017, build the solution and publish it. You could start the application in the 'Target directory' you have published to with:
```
Comment (barnstee) [inline README.md]: can, not could

File: README.md [hunk 15]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+To enable name resolution from within the container to the host network, you need to create a user define docker bridge network and connect the container to this network using the `--network`option as in this example:
+
+    docker network create -d bridge iot_edge
+    docker run --network iot_edge microsoft/iot-edge-opc-publisher <applicationname> [<iothubconnectionstring>] [options]
+
+### Access the container via a name on the network
+To access the Publisher container via a name from another container you need to assign a name to the container using the `--name` option as in this example:
+
+    docker run --name publisher microsoft/iot-edge-opc-publisher <applicationname> [<iothubconnectionstring>] [options]
+The Publisher container could now be reached by other containers via the name `publisher`over the network.
```
Comment (barnstee) [inline README.md]: can, not could

File: README.md [hunk 16]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+    docker run microsoft/iot-edge-opc-publisher <applicationname> [<iothubconnectionstring>] [options]
+
+## Important when using a container
+
+### Access to the Publisher OPC UA server
+The Publisher OPC UA server listens by default on port 62222. To expose this inbound port in a container, you need to use `docker run` option `-p`:
+
+    docker run -p 62222:62222 microsoft/iot-edge-opc-publisher <applicationname> [<iothubconnectionstring>] [options]
+
+### Access to the host network
```
Comment (barnstee) [inline README.md]: this isn't access to the host network, this is access to the publisher container network. I would not document this but actually describe how  --network=host works.

File: README.md [hunk 17]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+# Debugging the Application
+
+## Native on Winodws
+
+Open the OpcPublisher.sln project with Visual Studio 2017 and start debugging the app by hitting F5.
+
+## In a docker container
+
+Visual Studio 2017 supports debugging of application in docker container. This is done by using docker-compose. Since this does not allow to pass command line parameters it is not convenient. 
+Another debugging option VS2017 supports is to debug via ssh. In the root of the repository the docker build configuration file `Dockerfile.ssh`could be used to create a SSH enabled container by:
```
Comment (barnstee) [inline README.md]: can, not could

File: README.md [hunk 18]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
+          "request": "launch"
+        }
+      ]
+    }
+
+Build your project and publish it to a directory of your choice.
 
-## From Docker
-You can also run the application in a Docker container using the Dockerfile provided. From the root of the repo, in a console, type:
+Use a tool like WinSCP to copy over the published files to the container into the directory `/root/publisher` (this could be also a different directory, but needs to be in sync with the `cdw` property of launch.json.
```
Comment (barnstee) [inline README.md]: can, not could

File: README.md [hunk 19]
```diff
@@ -5,48 +5,452 @@ This reference implementation demonstrates how Azure IoT Edge can be used to con
...
 
-```docker build -t gw .```
+Now you could start debugging with the following command in Visual Studio's Command Window (View->Other Windows->Command Window):
```
Comment (barnstee) [inline README.md]: can, not could

File: src/OpcPublisher.csproj [hunk 20]
```diff
@@ -2,13 +2,13 @@
...
-    <AssemblyName>Opc.Ua.Publisher</AssemblyName>
+    <AssemblyName>OpcPublisher</AssemblyName>
     <OutputType>Exe</OutputType>
-    <PackageId>Opc.Ua.Publisher</PackageId>
+    <PackageId>OpcPublisher</PackageId>
     <RuntimeFrameworkVersion>1.1.2</RuntimeFrameworkVersion>
     <GenerateAssemblyCompanyAttribute>true</GenerateAssemblyCompanyAttribute>
     <GenerateAssemblyVersionAttribute>true</GenerateAssemblyVersionAttribute>
-    <Description />
+    <Description>This package subscribes to a configurable set of nodes in OPC UA server systems and publish them to Azure IoTHub.</Description>
```
Comment (barnstee) [inline src/OpcPublisher.csproj]: application, not package

File: src/OpcSession.cs [hunk 21]
```diff
@@ -205,184 +264,322 @@ public OpcSession(Uri endpointUri, uint sessionTimeout)
...
             try
             {
-                MonitoredItemInfo monitoredItemInfo = MonitoredItemsInfo.First(m => m.StartNodeId == nodeId);
+                if (PublisherShutdownInProgress)
+                {
+                    return;
+                }
+
+                // find a subscription we could the node monitor on
```
Comment (barnstee) [inline src/OpcSession.cs]: Grammar...

File: src/OpcStackConfiguration.cs [hunk 22]
```diff
@@ -0,0 +1,372 @@
...
+
+
+            //
+            // TransportConfigurations
+            //
+
+            Configuration.TransportQuotas = new TransportQuotas();
+            // the OperationTimeout should be twice the minimum value for PublishingInterval * KeepAliveCount, so set to 120s
+            Configuration.TransportQuotas.OperationTimeout = Program.OpcOperationTimeout;
+            Configuration.TransportQuotas.MaxStringLength = 1048576;
```
Comment (barnstee) [inline src/OpcStackConfiguration.cs]: Please reduce these again

File: src/Program.cs [hunk 23]
```diff
@@ -462,88 +485,116 @@ public static void Main(string[] args)
...
+                    foreach (var endpointUrl in uniqueEndpointUrls)
                     {
                         // create new session info.
-                        OpcSession opcSession = new OpcSession(endpointUri, OpcSessionCreationTimeout);
+                        OpcSession opcSession = new OpcSession(endpointUrl, OpcSessionCreationTimeout);
 
-                        // add monitored item info for all nodes to publish for this endpoint URI.
-                        var nodesOnEndpointUri = NodesToPublish.Where(n => n.EndPointUri.Equals(endpointUri));
-                        foreach (var nodeInfo in nodesOnEndpointUri)
+                        // create for all different OPC publishing intervals to this endpoint separate subscriptions
```
Comment (barnstee) [inline src/Program.cs]: Grammar...

File: src/PublisherServer.cs [hunk 24]
```diff
@@ -2,28 +2,133 @@
...
+                }
+            }
+
+            return resourceManager;
+        }
+
+        /// <summary>
+        /// Called after the server has been started.
+        /// </summary>
+        protected override void OnServerStarted(IServerInternal server)
```
Comment (barnstee) [inline src/PublisherServer.cs]: Please remove since this is not used (see implementation below).

**Review by barnstee — CHANGES_REQUESTED** [review]

File: src/IotHubMessaging.cs (hunk 1 above)
Comment (hansgschossmann) [reply src/IotHubMessaging.cs]: fixed #Resolved

File: src/Namespaces.cs (hunk 2 above)
Comment (hansgschossmann) [reply src/Namespaces.cs]: will change to: http://microsoft.com/Opc/Publisher/ #Resolved

File: src/OpcSession.cs (hunk 4 above)
Comment (hansgschossmann) [reply src/OpcSession.cs]: great catch. thx

File: src/OpcSession.cs (hunk 3 above)
Comment (hansgschossmann) [reply src/OpcSession.cs]: yes that makes sense. fixed

File: src/OpcSession.cs (hunk 5 above)
Comment (hansgschossmann) [reply src/OpcSession.cs]: yes that makes sense. fixed #Resolved

File: Dockerfile (hunk 6 above)
Comment (hansgschossmann) [reply Dockerfile:9]: it works

File: README.md (hunk 7 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 8 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 9 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 10 above)
Comment (hansgschossmann) [reply README.md]: fxied

File: README.md (hunk 11 above)
Comment (hansgschossmann) [reply README.md]: changed to a complete new format

File: README.md (hunk 12 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 13 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 14 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 16 above)
Comment (hansgschossmann) [reply README.md]: host network documentation is sparse, almost not there. but you are right it does not enable host network access, but inter container name resolution. changed

File: README.md (hunk 15 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 17 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: README.md (hunk 19 above)
Comment (hansgschossmann) [reply README.md]: fixed

File: src/OpcSession.cs (hunk 21 above)
Comment (hansgschossmann) [reply src/OpcSession.cs]: fixed

File: src/PublisherServer.cs (hunk 24 above)
Comment (hansgschossmann) [reply src/PublisherServer.cs]: removed

File: src/Program.cs (hunk 23 above)
Comment (hansgschossmann) [reply src/Program.cs]: fixed

File: src/OpcStackConfiguration.cs (hunk 22 above)
Comment (hansgschossmann) [reply src/OpcStackConfiguration.cs]: done

File: src/OpcPublisher.csproj (hunk 20 above)
Comment (hansgschossmann) [reply src/OpcPublisher.csproj]: changed

File: src/Namespaces.cs (hunk 2 above)
Comment (mregen) [reply src/Namespaces.cs]: I checked the spec, (Part 6 F14 has a sample) it should be ok like this

---
In reply to: [141532429](https://github.com/Azure/iot-edge-opc-publisher/pull/78#discussion_r141532429) [](ancestors = 141532429)

File: README.md [hunk 25]
```diff
@@ -7,11 +7,11 @@ This application, apart from including an OPC UA *client* for connecting to exis
 
 The application is implemented using .NET Core technology and is able to run on the platforms supported by .NET Core.
 
-Publisher implements a retry logic to establish connections to endpoints which have not responded to a certain number of keep alive requests, if the OPC UA server on this endpoint had a power outage.
+Publisher implements a retry logic to establish connections to endpoints which have not responded to a certain number of keep alive requests, for example if the OPC UA server on this endpoint had a power outage.
 
-For each distinct publishing interval to a OPC UA server it creates a separate subscription over which all nodes with this publishing interval are updated.
+For each distinct publishing interval to aanOPC UA server it creates a separate subscription over which all nodes with this publishing interval are updated.
```
Comment (barnstee) [inline README.md]: typo

**Review by barnstee — APPROVED** [review]

File: Dockerfile (hunk 6 above)
Comment (barnstee) [reply Dockerfile:9]: ok

File: README.md (hunk 25 above)
Comment (hansgschossmann) [reply README.md]: fixed

**Pull request merged**